imap_processing/lo/l1b/lo_l1b.py

```diff
@@ -418,76 +418,157 @@ def set_spin_cycle_from_spin_data(
             )
         spin_cycle = np.array(spin_cycle)
-        l1b_science["spin_cycle"] = xr.DataArray(
-            spin_cycle,
-            dims=["epoch"],
-        )
-    elif l1a_science.attrs["Logical_source"] == "imap_lo_l1a_histogram":
-        # For histogram: keep 2D array (n_epochs, 7)
-        esa_steps = l1b_science["esa_step"].values  # shape: (7,)
+        l1b_valid["spin_cycle"] = xr.DataArray(spin_cycle, dims=["epoch"])
+    elif logical_src == "imap_lo_l1a_histogram":
+        # For histogram: keep 2D array (n_valid_epochs, esa_step)
+        esa_steps = l1b_valid["esa_step"].values  # shape: (7,)
         spin_cycle = spin_start_num_per_asc + 7 + (esa_steps - 1) * 2
-        l1b_science["spin_cycle"] = xr.DataArray(
-            spin_cycle,
-            dims=["epoch", "esa_step"],
-        )
+        l1b_valid["spin_cycle"] = xr.DataArray(spin_cycle, dims=["epoch", "esa_step"])
     else:
-        raise ValueError("set spin cycle called with unsupported dataset")
+        raise ValueError(
+            "set spin cycle called with unsupported dataset with "
+            "Logical_source: {logical_src}"
+        )
 
-    return l1b_science
+    return l1b_valid
+
+
+def match_science_to_spin_asc(
+    science_met_per_asc: xr.DataArray, spin_met_per_asc: xr.DataArray
+) -> tuple[np.ndarray, np.ndarray]:
+    """
+    Compute the indices of the closest spin acquisition times for each science event.
+
+    This function matches science data acquisition epochs to spin data acquisition
+    epochs by finding the closest spin acquisition indices for each science data
+    acquisition epoch. The result is an array where each element corresponds to the
+    index of the closest spin data acquisition time for a given science event.
+
+    Parameters
+    ----------
+    science_met_per_asc : xr.DataArray
+        An array of science acquisition epochs in MET seconds.
+    spin_met_per_asc : xr.DataArray
+        An array of spin acquisition epochs in MET seconds.
+
+    Returns
+    -------
+    closest_start_acq_indices : np.ndarray
+        An array of indices representing the closest spin acquisition times for each
+        science event.
+    time_diff : np.ndarray
+        An array of time differences between each science event and all spin acquisition
+        times.
+    """
+    # Find the closest start_acq for each direct event
+    # computes the index of the closest spin_met_per_asc for each science_met_per_asc
+    # so the resulting array will be of length len(science_met_per_asc), one index per
+    # ASC, but the value of each index will be the index of the closest spin data.
+    # Find the index of the spin data that starts just before or at each science epoch
+    time_diff = science_met_per_asc[:, None] - spin_met_per_asc
+    time_diff_masked = np.where(time_diff >= 0, time_diff, np.inf)
+    closest_start_acq_indices = time_diff_masked.argmin(axis=1)
+
+    return closest_start_acq_indices, time_diff
```

Comment:
Do you need to return time_diff here, or can you set all time_diff < 0 to an index of -1 or something that indicates it is a bad index for use later instead?